ddtrace/internal/datadog/profiling/dd_wrapper/test/forking.cpp

```diff
@@ -0,0 +1,150 @@
+#include "interface.hpp"
+#include "test_utils.hpp"
+#include <gtest/gtest.h>
+
+#include <chrono>
+
+// Initiate an upload in a separate thread, otherwise we won't be mid-upload during fork
+void
+upload_in_thread()
+{
+    std::thread([&]() { ddup_upload(); }).detach();
+}
+
+[[noreturn]] void
+profile_in_child(unsigned int num_threads, unsigned int run_time_ns, std::atomic<bool>& done)
+{
+    // Assumes setup has been called. Launch some samplers, wait, upload, return.
+    // Exit because we don't want this process to exit
+    std::vector<unsigned int> ids;
+    for (unsigned int i = 0; i < num_threads; i++) {
+        ids.push_back(i);
+    }
+    std::vector<std::thread> new_threads;
+    launch_samplers(ids, 10e3, new_threads, done);
+    std::this_thread::sleep_for(std::chrono::nanoseconds(run_time_ns));
+    done.store(true);
+    ddup_upload();
+    std::exit(0);
+}
+
+bool
+is_exit_normal(int status)
+{
+    return WIFEXITED(status) && WEXITSTATUS(status) == 0;
+}
+
+// Validates that sampling/uploads work around forks
+void
+sample_in_threads_and_fork(unsigned int num_threads, unsigned int sleep_time_ns)
+{
+    configure("my_test_service", "my_test_env", "0.0.1", "https://localhost:8126", "cpython", "3.10.6", "3.100", 256);
+    std::atomic<bool> done(false);
+    std::vector<std::thread> threads;
+    std::vector<unsigned int> ids;
+    for (unsigned int i = 0; i < num_threads; i++) {
+        ids.push_back(i);
+    }
+
+    // ddup is configured, launch threads
+    launch_samplers(ids, sleep_time_ns, threads, done);
+
+    // Collect some profiling data for a few ms, then upload in a thread before forking
+    std::this_thread::sleep_for(std::chrono::milliseconds(500));
+    upload_in_thread();
+
+    // Fork, wait in the parent for child to finish
+    pid_t pid = fork();
+    if (pid == 0) {
+        // Child
+        profile_in_child(num_threads, 500e3, done); // Child profiles for 500ms
```

Comment:
It took me a second to realize that this is `noreturn`, rather than an exponential number of forks. 🍴 💣

Reply:
I did consider the possibility of using a fork-bomb type strategy to get _lots_ of different tests running simultaneously, but I decided against it.